### ISSUE TITLE
Optimise threshold for MeshXLinear

### ISSUE THREAD

**Issue by nsorros** [wellcometrust/grants_tagger#112]
In multilabel classification where you can have more than one labels assigned to an example, it is typically to have a threshold to control which labels are assigned with the default often being 50%. As you have different examples for each label, a custom threshold per label tends to perform better as this [paper](https://pralab.diee.unica.it/sites/default/files/pillai_PR2013_Thresholding_0.pdf) suggests.

I have implemented it [here](https://github.com/wellcometrust/grants_tagger/blob/master/grants_tagger/tune_threshold.py) but there are some problems I am facing when using the code
* it takes a lot of time to run
* it overfits
* memory requirements are big

In regards to memory requirements and potentially time, this can be improved by using sparse probabilities i.e. zero ing probs less than a value, say 0.01 and then using sparse matrices. 

In regards to overfit, we should use a substantially big validation set say 100K as they do in the literature and keep an equal or smaller test set, say 20K to 100K. As context, I had been using a test set of 1.3M and using 100K for tuning the thresholds which I was seeing that it was overfitting to the examples it was seeing.
